modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/DownloadSpreadsheetButton/DownloadSpreadsheetButton.js

```diff
@@ -0,0 +1,313 @@
+
+			setTimeout(() => {
+				onClose();
+				Liferay.Util.navigate(tentativeNavigationPath);
+			}, DEFAULT_TIMEOUT_DELAY + FEEDBACK_TIMEOUT_DELAY);
+		},
+		[tentativeNavigationPath]
+	);
+
+	const preventHardReloadWhileGenerating = useCallback(
+		(event) => {
+			if (!loading) {
+				return;
+			}
+			event.preventDefault();
+			event.returnValue = '';
+		},
+		[loading]
+	);
+
+	useEffect(() => {
+		const navigationEventHandler = Liferay.on('beforeNavigate', (event) => {
+			if (!loading) {
+				return;
+			}
+
+			Liferay.fire('closeApplicationsMenu');
+			setShowPendingRequestModal(true);
+			setTentativeNavigationPath(event.path);
+			event.originalEvent.preventDefault();
+		});
+
+		window.addEventListener(
+			'beforeunload',
+			preventHardReloadWhileGenerating
+		);
+
+		return () => {
+			clearTimeout(defaultDelayTimeout);
+			clearTimeout(feedbackDelayTimeout);
+			navigationEventHandler.detach();
+			window.removeEventListener(
+				'beforeunload',
+				preventHardReloadWhileGenerating
+			);
+		};
+	}, [
+		defaultDelayTimeout,
+		feedbackDelayTimeout,
+		loading,
+		preventHardReloadWhileGenerating,
+	]);
+
+	return (
+		<>
+			<ClayTooltipProvider>
+				<ClayButton
+					borderless
+					className={classnames('download-spreadsheet-button', {
+						'download-spreadsheet-button--loading': loading,
+						'text-danger':
+							feedbackStatus.show &&
+							feedbackStatus.type === 'danger',
+						'text-success':
+							feedbackStatus.show &&
+							feedbackStatus.type === 'success',
+						'text-warning':
+							feedbackStatus.show &&
+							feedbackStatus.type === 'warning',
+					})}
+					data-tooltip-align="top"
+					disabled={disabledGenerateButton}
+					displayType="secondary"
+					onClick={handleClick}
+					title={Liferay.Language.get(
+						'download-your-data-in-a-xls-file'
+					)}
+				>
+					<span className="inline-item inline-item-before">
+						{loading ? (
+							<ClayLoadingIndicator small />
+						) : (
+							<ClayIcon symbol={clayIconSymbol} />
+						)}
+					</span>
+					{buttonTextKey}
+				</ClayButton>
+			</ClayTooltipProvider>
+
+			{loading && (
+				<ClayTooltipProvider>
+					<ClayButtonWithIcon
+						borderless
+						className="ml-2"
+						data-tooltip-align="top"
+						displayType="secondary"
+						onClick={handleCancelRequest}
+						symbol="times-circle"
+						title={Liferay.Language.get('cancel-export')}
+					/>
+				</ClayTooltipProvider>
+			)}
+
+			{toastMessage.show && (
+				<ClayAlert.ToastContainer>
+					<ClayAlert
+						autoClose={5000}
+						className="download-spreadsheet-button__alert"
+						displayType={toastMessage.type}
+						onClose={handleToastClose}
+					>
+						{toastMessage.content}
+					</ClayAlert>
+				</ClayAlert.ToastContainer>
+			)}
+
+			<DownloadSpreadsheetModal
+				disableSecondaryButton={disableCancelAndLeaveButton}
+				secondaryButtonClickCallback={secondaryClickModalHandler}
+				secondaryButtonText={cancelAndLeaveButtonText}
```

Comment:
I was thinking that maybe it would be clearer if we move part of the state of the Button to the Modal. I see that cancelAndLeaveButtonText and disableCancelAndLeaveButton is more a responsibility of the modal component. Does it makes sense to you?